### DISCUSSION TITLE
Should Micronaut framework 5 drop Java 17?

### DISCUSSION THREAD

**Discussion by sdelamo** [micronaut-projects/micronaut-core#10484] [Polls]
Should we set Micronaut Framework 5 Java baseline as Java 21?

see: https://github.com/micronaut-projects/micronaut-core/issues/10483

**Comment by nedelva** [comment]
I am all for progress and moving forward to using Java 21 already, and the arguments for doing so are quite compelling. That being said, I would be very cautious about dropping support for Java 17. This kind of move could potentially scare away a lot of folks who thrust to have some sort of stability in using the combo Java 17+ Micronaut.

If you want to do it at least gather a larger consensus from the user base; a poll here with just a handful of votes is not going to suffice for that.

**Reply by sdelamo** [reply]
how would you gather that consensus from the user base?

**Reply by nedelva** [reply]
Surely there are some Micronaut social media accounts out there for public announcements? Like Twitter or the project's blog?

**Reply by sgammon** [reply]
Personally, I think a poll here is fine; reasoning that Micronaut 5 = Java 21+, and Micronaut 4 = Java 17+, makes a lot of sense. These are LTS releases and major version releases, after all.

**Comment by yawkat** [comment]
Dropping support for pre-21 would simplify some virtual thread handling. Other than that, I don't think there are big advantages.

Migration from 17 to 21 seems to be relatively painless for users, so there is less reason to stay on 17.

Unless there are other big upsides to 21 than virtual threads, I don't think it is worth forcing users to update, even if it would be painless for them.

**Reply by sdelamo** [reply]
I have summarized the advantages I see here: https://github.com/micronaut-projects/micronaut-core/issues/10483

**Comment by sgammon** [comment]
Java 22 support would gain a lot: FFM, scoped values, stream gatherers, and so on. I wish these had landed in Java 21 but no such luck. One other consideration would definitely be [JPMS](https://github.com/micronaut-projects/micronaut-core/issues/6395). We are using Micronaut with GraalVM, and GVM has already moved to using the `modulepath` for certain features or dependencies, which is complicating our use of Micronaut.

In general it is going to be interesting to see how Netty evolves in the presence of new support like Virtual Threads and FFM. Java 21 would help get everyone there even if Java 22 is a little further out.

Obviously, Gradle plugins will still need to run under Java 11 or 17, but that's another story.

**Reply by yawkat** [reply]
These are user-facing features but not particularly useful for micronaut the framework. Users can use 22+ even if we stay on 17.

Netty will not move past 8 baseline in the near future.

**Comment by sdelamo** [comment]
The core committers teams, Unity Foundation + Oracle Labs, had a Micronaut Framework 5 planning meeting last Thursday. We have decided to keep the Java 17 baseline for Micronaut Framework 5.
